cmd/dagger/engine.go

```diff
@@ -71,141 +52,38 @@ func withEngineAndTUI(
 
 	params.DisableHostRW = disableHostRW
 
-	if params.JournalFile == "" {
-		params.JournalFile = os.Getenv("_EXPERIMENTAL_DAGGER_JOURNAL")
-	}
+	params.EngineNameCallback = Frontend.ConnectedToEngine
 
-	if interactive {
-		return interactiveTUI(ctx, params, fn)
-	}
-	if useLegacyTUI {
-		if progress == "auto" && autoTTY || progress == "tty" {
-			return legacyTUI(ctx, params, fn)
-		} else {
-			return plainConsole(ctx, params, fn)
-		}
-	}
-	return runWithFrontend(ctx, params, fn)
-}
+	params.CloudURLCallback = Frontend.ConnectedToCloud
 
-// TODO remove when legacy TUI is no longer supported; this has been
-// assimilated into idtui.Frontend
-func plainConsole(ctx context.Context, params client.Params, fn runClientCallback) error {
-	opts := []console.WriterOpt{
-		console.ShowInternal(debug),
-	}
-	if debug {
-		opts = append(opts, console.WithMessageLevel(progrock.MessageLevel_DEBUG))
-	}
-	progW := console.NewWriter(os.Stderr, opts...)
-	params.ProgrockWriter = progW
-	params.EngineNameCallback = func(name string) {
-		fmt.Fprintln(os.Stderr, "Connected to engine", name)
-	}
-	params.CloudURLCallback = func(cloudURL string) {
-		fmt.Fprintln(os.Stderr, "Dagger Cloud URL:", cloudURL)
-	}
-	engineClient, ctx, err := client.Connect(ctx, params)
-	if err != nil {
-		return err
-	}
-	defer engineClient.Close()
-	return fn(ctx, engineClient)
-}
+	params.EngineTrace = Frontend
+	params.EngineLogs = Frontend
 
-func runWithFrontend(
-	ctx context.Context,
-	params client.Params,
-	fn runClientCallback,
-) error {
-	frontend := idtui.New()
-	frontend.Debug = debug
-	frontend.Plain = progress == "plain"
-	frontend.Silent = silent
-	params.ProgrockWriter = frontend
-	params.EngineNameCallback = frontend.ConnectedToEngine
-	params.CloudURLCallback = frontend.ConnectedToCloud
-	return frontend.Run(ctx, func(ctx context.Context) error {
-		sess, ctx, err := client.Connect(ctx, params)
-		if err != nil {
-			return err
+	if exp, ok := telemetry.ConfiguredSpanExporter(ctx); ok {
+		if !telemetry.ForceLiveTrace {
+			exp = telemetry.FilterLiveSpansExporter{
+				// SpanProcessor: processor,
```

Comment:
Yeah, man, honestly it's very dense but I can't say it's without merit; instrumenting the world in a way that's compatible with all these use cases is just very complicated, and even with all the existing complexity we've had to do funky things of our own, like having a "live span processor" that accepts in-flight spans and sends them to the exporter while they're running, whereas normally spans are only exported when they complete.

For me it's mostly cobbled together knowledge + a long ongoing chat with ChatGPT, but maybe the [Glossary](https://opentelemetry.io/docs/concepts/glossary/) would help?

tl;dr for some terms you can expect to see:

* resource: roughly represents a component, e.g. the running process; has typed attributes, usually just a handful of these
* span: probably familiar already but including for completeness; measures a span of time spent doing some task, associated to a resource (analogous to a component), can have arbitrary typed attributes and events, which are all held in memory (so don't use those for e.g. logs)
* log record: a chunk of output, but it's actually a typed value, and also has attributes, is tied to a resource, etc.
* [log|span] processor: accepts logs and spans as they appear and passes them along to an exporter
* [log|spab] exporter: sends a batch of spans to a consumer, e.g. an OTLP endpoint, our CLI frontend, or Cloud
* gRPC unary interceptor: instruments simple gRPC call/response APIs with OpenTelemetry, has both client and server hooks
* gRPC stream interceptor: instruments streaming gRPC call/response APIs with OpenTelemetry, has both client and server hooks
* TracerProvider: basically a factory for Tracers, typically globally configured, since it's what has exporters/span processors configured inside and you don't want to have to DI those everywhere
* Tracer: a factory for spans, initialized with a component name (I've made this align roughly with Go packages, e.g. `dagger.io/core`, which isn't a real Go package, but it hasn't mattered yet)

Drawing the line there for now. :sweat_smile: Feel free to ask anythign else.

And just to reiterate since we're doing something weird: our "catch" is that our CLI and Cloud accept "live" spans, whereas most OpenTelemetry systems (e.g. Honeycomb) only accept completed spans, and will get very confused if you send them "live" spans.

Reply:
Yeah, it's pretty cobbled together. I guess for every day development, the tracer/provider/factories yadda yadda don't really matter.

The gist of it is OTEL is actually 3 projects:
- Tracing: which we use extensively
- Logs: which are pretty simple compared to Tracing, and are still experimental
- Metrics: which we are NOT using yet -- but this could be super interesting. @sipsma For instance, the engine logs you added to dump memory/cpu/etc could now use OTEL metrics. We could also push this even further by exporting container-level metrics and graphing that in Cloud

For tracing, the gist of the terminology is here: https://opentelemetry.io/docs/specs/otel/overview/#traces

Translation manual:

- `Trace` == `Run`. A Trace is just a `TraceID` used to tie `Span`s together, exactly like `Run`/`RunID` and vertices.
- `Span` == `Vertex-ish`. It's got a start time, end time, state, and arbitrary key/values attached. It can also have "events" (message + key/value at a point in time during the span) -- I guess similar to `VertexStatuses`?
- `Resource` == `Label` (?). Key/value map attached to all spans. Conventionally in tracing this contains stuff like service name, PID, etc. We are stashing the "engine labels" in there (git information, etc)